proof/refine/ARM/SchedContextInv_R.thy

```diff
@@ -1012,6 +1012,289 @@ lemma refillUpdate_corres:
    apply clarsimp+
   done
 
+lemma head_insufficient_length_greater_than_one:
+  "\<lbrakk>the (head_insufficient sc_ptr s);
+    pred_map (\<lambda>cfg. unat MIN_BUDGET \<le> refills_unat_sum (scrc_refills cfg)) (sc_refill_cfgs_of s) sc_ptr\<rbrakk>
+   \<Longrightarrow> pred_map (\<lambda>cfg. Suc 0 < length (scrc_refills cfg)) (sc_refill_cfgs_of s) sc_ptr"
+  apply (frule head_insufficient_true_imp_insufficient)
+   apply (clarsimp simp: vs_all_heap_simps)
+  apply (clarsimp simp: vs_all_heap_simps sc_at_ppred_def refills_unat_sum_def word_less_nat_alt)
+  apply (case_tac "sc_refills y"; fastforce dest!: member_le_sum_list)
+  done
+
+lemma length_sc_refills_cross:
+  "\<lbrakk>(s, s') \<in> state_relation; sc_at scp s; sc_at' scp s'; valid_refills' scp s'\<rbrakk>
+   \<Longrightarrow> ((\<lambda>sc. P (length (sc_refills sc))) |< scs_of2 s) scp
+        = ((\<lambda>sc'. P (scRefillCount sc')) |< scs_of' s') scp"
+  apply (drule (2) state_relation_sc_relation)
+  apply (clarsimp simp: obj_at_simps is_sc_obj valid_refills'_def sc_relation_def opt_map_red)
+  done
+
+lemma head_insufficient_length_cross:
```

Comment:
I think you can now remove this lemma. It is not used and we probably don't want it in.